Упражнение 3

seal-developer opened the pull request.

seal-developer committed:
Упражнение 3

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.css:
```diff
@@ -0,0 +1,245 @@
+  text-align: center;
+}
+
+input[type='radio']:focus,
+input[type='radio']:focus-visible {
```
Лучше классами делать первую часть, так как это сильно общий селектор

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.css:
```diff
@@ -0,0 +1,245 @@
+  outline-style: solid;
+}
+
+#form_page h2 {
```
h2 можно поменять на класс

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.css:
```diff
@@ -0,0 +1,245 @@
+  flex-wrap: wrap;
+}
+
+.block_results h3 {
```
h3 можно поменять на класс

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.css:
```diff
@@ -0,0 +1,245 @@
+}
+
+.error {
+  color: rgb(255, 255, 255);
```
А почему здесь `rgb`, а обычно hex?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.css:
```diff
@@ -0,0 +1,245 @@
+  color: rgb(255, 255, 255);
+}
+
+.control label {
```
label можно поменять на класс

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.css:
```diff
@@ -0,0 +1,245 @@
+  width: 20px;
+  height: 20px;
+  display: inline-flex;
+  justify-content: center;
+  align-items: center;
+  vertical-align: middle;
```
Нужны ли все эти параметры вместе?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.css:
```diff
@@ -0,0 +1,245 @@
+  margin: 0 0 30px;
+}
+
+input[type='radio'] {
```
Опять же, лучше на класс, и лучше такие вещи держать вместе, чуть более очевидно.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.css:
```diff
@@ -0,0 +1,245 @@
+  color: #25ffe0;
+}
+
+#checkboxList label {
```
label лучше на класс

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,116 @@
+      checkboxList.innerHTML = '';
+    }
+
+    const errorBoxClassName = '.error--box';
```
Можно убрать наверх

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,116 @@
+    const enteredValues = document.getElementById(enteredValuesID).value;
+    const arrayEnteredValues = enteredValues.split(' ');
+
+    if (arrayEnteredValues.length >= 20) {
```
С таким большим количеством вложенных `if` тяжело что-то понимать, так что это надо убирать и разбивать на мелкие функции.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,116 @@
+    const arrayEnteredValues = enteredValues.split(' ');
+
+    if (arrayEnteredValues.length >= 20) {
+      const error = document.createElement('span');
+      error.className = 'error';
+      error.innerHTML =
+        'Вы ввели недопустимое количество чисел! Введите не более 20 чисел.';
+      errorBox.append(error);
```
Можно вынести здание ошибки в отдельную функцию.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,116 @@
+        'Вы ввели недопустимое количество чисел! Введите не более 20 чисел.';
+      errorBox.append(error);
+    } else {
+      for (let enteredValue of arrayEnteredValues) {
```
`let`->`const`

Tauders replied:
Можно валидацию каждой числа убрать в функцию

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,116 @@
+    const enteredValues = document.getElementById(enteredValuesID).value;
+    const arrayEnteredValues = enteredValues.split(' ');
+
+    if (arrayEnteredValues.length >= 20) {
+      const error = document.createElement('span');
+      error.className = 'error';
+      error.innerHTML =
+        'Вы ввели недопустимое количество чисел! Введите не более 20 чисел.';
+      errorBox.append(error);
+    } else {
```
Можно убрать в функцию

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,116 @@
+          errorBox.append(error);
+          checkboxList.innerHTML = '';
+          break;
+        } else if (enteredValue === '') {
+          continue;
```
Можно заранее убрать пустые значения

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,116 @@
+          const checkboxItem = document.createElement('input');
+
+          checkboxItem.setAttribute('type', 'checkbox');
+          checkboxItem.className = 'checkbox';
```
Можно убрать в константу

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,116 @@
+          }
+
+          const labelItem = document.createElement('label');
+          const checkboxItem = document.createElement('input');
+
+          checkboxItem.setAttribute('type', 'checkbox');
+          checkboxItem.className = 'checkbox';
+          checkboxItem.id = enteredValue;
```
Убрать генерацию элементов можно в функцию

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,116 @@
+  .addEventListener('submit', function (e) {
+    e.preventDefault();
+
+    const radioAdditionOption = document.getElementById('addition');
+    const radioSubtractionOption = document.getElementById('subtraction');
+    const radioMultiplicationOption = document.getElementById('multiplication');
+    const radioDivisionOption = document.getElementById('division');
+    const checkboxElementsClassName = '#checkboxList label input';
```
Можно убрать в константы

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,116 @@
+      checkboxElementsClassName
+    );
+    const arrayEnteredValuesChecked = [];
+    for (let i = 0; i < checkboxElements.length; i++) {
+      if (checkboxElements[i].checked) {
+        arrayEnteredValuesChecked.push(+checkboxElements[i].id);
+      }
+    }
```
Почему бы не `for ... of`?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,116 @@
+      }
+    }
+
+    if (radioAdditionOption.checked) {
+      countingResults.innerHTML = arrayEnteredValuesChecked.reduce(
+        (addition, current) => addition + current
+      );
+
+      radioAdditionOption.checked = false;
+    } else if (radioSubtractionOption.checked) {
+      countingResults.innerHTML = arrayEnteredValuesChecked.reduce(
+        (subtraction, current) => subtraction - current
+      );
+      radioSubtractionOption.checked = false;
+    } else if (radioMultiplicationOption.checked) {
+      countingResults.innerHTML = arrayEnteredValuesChecked.reduce(
+        (multiplication, current) => multiplication * current
+      );
+      radioMultiplicationOption.checked = false;
+    } else if (radioDivisionOption.checked) {
+      countingResults.innerHTML = arrayEnteredValuesChecked.reduce(
+        (division, current) => division / current
+      );
+      radioDivisionOption.checked = false;
+    }
```
Можно оставить здесь только вычисление, а присвоение и сброс радио батонов после.

Tauders replied:
Можно убрать вычисление убрать в отдельную функцию.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,116 @@
+      radioMultiplicationOption.checked = false;
+    } else if (radioDivisionOption.checked) {
+      countingResults.innerHTML = arrayEnteredValuesChecked.reduce(
+        (division, current) => division / current
```
А если на 0?

seal-developer committed:
Правки

seal-developer committed:
1. Объединила функции ошибок в одну с входным параметром; 2. Поправила ошибку при делении на ноль, чтобы не отображалось "undefined" в блоке результата; 3. Перенесла функции; 4. Поправила опечатку (перепутала функции при правках)

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,159 @@
+    );
+    const radioDivisionOption = document.getElementById(radioDivisionOptionID);
+
+    function resetRadioButtons() {
+      radioAdditionOption.checked = false;
+      radioSubtractionOption.checked = false;
+      radioMultiplicationOption.checked = false;
+      radioDivisionOption.checked = false;
+    }
```
Надо вынести за пределы функции

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,159 @@
+
+    function performСalculationOperation(error) {
+      if (radioAdditionOption.checked) {
+        return arrayEnteredValuesChecked.reduce(function (addition, current) {
+          return addition + current;
+        });
```
Можно чуть проще написать через стрелочную функцию, попробуй.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,159 @@
+      radioDivisionOption.checked = false;
+    }
+
+    function performСalculationOperation(error) {
```
Вынеси за пределы функции

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,159 @@
+      } else if (radioDivisionOption.checked) {
+        if (arrayEnteredValuesChecked.includes(0)) {
+          createError('Делить на ноль нельзя!');
+          return (error = '');
```
А это зачем?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,159 @@
+        return arrayEnteredValuesChecked.reduce(function (addition, current) {
+          return addition + current;
+        });
+      } else if (radioSubtractionOption.checked) {
```
А нужно `else if`?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,159 @@
+    } else {
+      function CheckEnteredValues() {
+        for (const enteredValue of arrayEnteredValues) {
+          if (enteredValue === '') {
+            continue;
+          } else if (isNaN(enteredValue)) {
+            createError('Вы ввели нечисловое значение! Попробуйте снова.');
+            break;
+          } else if (enteredValue.length >= 20) {
+            createError(
+              'Вы ввели слишком длинное число! Максимально разрешенная  длина числа - 20 знаков.'
+            );
+            break;
+          } else {
```
А нужно ли здесь `if ... else`? если ты каждое выражение заканчиваешь `break;` или `continue;`?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,159 @@
+        'Вы ввели недопустимое количество чисел! Введите не более 20 чисел.'
+      );
+    } else {
+      function CheckEnteredValues() {
```
Вынеси функцию наружу

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,159 @@
+            break;
+          } else {
+            clearErrorboxSheet();
+            function createCheckboxItem() {
```
Вынеси функцию наружу

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,159 @@
+  .addEventListener('submit', function (e) {
+    e.preventDefault();
+
+    const checkboxElementsClassName = '#checkboxList label input';
```
А почему имя класса начинается `#`?

seal-developer committed:
1. Вынесла функции; 2. Правка опечаток;
3. Поправила значение переменной;
4. Переделала логические конструкции;
5. Заменила функции операций на стрелочные.
6.  Расширила обработку ошибок.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -45,115 +48,152 @@ document
 
+    function CheckEnteredValues() {
+      for (const enteredValue of arrayEnteredValues) {
+        function createCheckboxItem() {
+          const labelItem = document.createElement('label');
+          const checkboxItem = document.createElement('input');
+
+          checkboxItem.setAttribute('type', 'checkbox');
+          checkboxItem.className = checkboxItemClassName;
+          labelItem.className = labelItemClassName;
+          checkboxItem.id = enteredValue;
+          labelItem.setAttribute('for', enteredValue);
+          checkboxList.append(labelItem);
+          labelItem.append(checkboxItem);
+          labelItem.append(enteredValue);
+        }
```
Вынеси эту функцию ещё выше.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,199 @@
+          checkboxItem.setAttribute('type', 'checkbox');
+          checkboxItem.className = checkboxItemClassName;
+          labelItem.className = labelItemClassName;
+          checkboxItem.id = enteredValue;
```
зачем ему такое id?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,199 @@
+    return createError('Для выполнения операций выберите числа!');
+  }
+
+  function addition() {
+    countingResults.innerHTML = arrayEnteredValuesChecked.reduce(
+      (addition, current) => addition + current
+    );
+  }
+  function subtraction() {
+    countingResults.innerHTML = arrayEnteredValuesChecked.reduce(
+      (subtraction, current) => subtraction - current
+    );
+  }
+  function multiplication() {
+    countingResults.innerHTML = arrayEnteredValuesChecked.reduce(
+      (multiplication, current) => multiplication * current
+    );
+  }
+  function division() {
```
Вынеси эти функции выше, добавь параметры массива для вычисления, пусть возвращает число.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,199 @@
+    const enteredValues = document.getElementById(enteredValuesID).value;
+    const arrayEnteredValues = enteredValues.split(' ');
+
+    function CheckEnteredValues() {
```
Вынеси функцию выше. Пусть она занимается только валидацией, то есть если всё хорошо, она вернёт `true`, если нет, ошибку. 
Почему эта функция с большое буквы?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,199 @@
+    const arrayEnteredValues = enteredValues.split(' ');
+
+    function CheckEnteredValues() {
+      for (const enteredValue of arrayEnteredValues) {
```
Создай функцию валидация каждого значения.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,199 @@
+      }
+    }
+
+    if (arrayEnteredValues.length >= 20) {
```
Эту проверку можно убрать в `CheckEnteredValues`

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,199 @@
+    document.getElementById(enteredValuesID).value = '';
+  });
+
+const radioAdditionOption = document.getElementById(radioAdditionOptionID);
+const radioSubtractionOption = document.getElementById(
+  radioSubtractionOptionID
+);
+const radioMultiplicationOption = document.getElementById(
+  radioMultiplicationOptionID
+);
+const radioDivisionOption = document.getElementById(radioDivisionOptionID);
```
Зачем это снаружи функции?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,199 @@
+  const controlRadioOptionIndex = defineIndex();
+
+  switch (controlRadioOptionIndex) {
+    case 0:
+      addition();
+      break;
+    case 1:
+      subtraction();
+      break;
+    case 2:
+      multiplication();
+      break;
+    case 3:
```
0, 1, 2, 3 плохие названия для определения, что ты хочешь сделать, они называются "магические числа" https://ru.wikipedia.org/wiki/Магическое_число_(программирование)
Сделай константы с определением, что они значат. 
Или ещё лучше, пусть возвращается более осмысленное название.

seal-developer replied:
Это индекс.  я вверху передаю переменную, которая его содержит, 1,2,3 - это значение переменной

Tauders replied:
Я понимаю, но даже так это плохие названия. Можно либо вынести индекс в переменные, чтобы было понятно, либо (лучше) возвращать не индекс, а конкретные действия. 

```
const ADDITIONAL_INDEX_ACTION = 0;
...

case ADDITIONAL_INDEX_ACTION:
```

Название можно поменять. 

А если бы был лучший вариант, то была бы функция, которая возвращает одно из действий уже. 

```
const action = getNeededAction();

switch (action) {
  case 'addition': 
  {
...
```

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,199 @@
+
+  clearErrorboxSheet();
+
+  const controlRadioOptionIndex = defineIndex();
```
`defineIndex` означает "определить индекс", а здесь ты не то делаешь, не очень хорошее название

seal-developer replied:
Он определяет индекс кнопки, что значит, не то делаю?

Tauders replied:
Он его не определяет, он его находит или берёт у радио батона. Define не значит взять готовое, а значит задать что-то.

seal-developer committed:
1. Вынесла функции; 2. Убрала id у чекбоксов, для функций использовала value;
3. Поправила опечатку;
4. Оставила в функции checkEnteredValues()
только валидациию;
5.Создала функцию валидация каждого значения;
6. Передела case у switch и функцию определяющую значение case.
7.Убрала в функцию объявление переменных, отвечающих за радио.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,211 @@
+    const enteredValues = document.getElementById(enteredValuesID).value;
+    const arrayEnteredValues = enteredValues.split(' ');
+
+    function createCheckboxItem(enteredValue) {
```
Всё ещё вынеси выше функцию, за пределы этого `eventListener`

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,211 @@
+      labelItem.append(checkboxItem);
+      labelItem.append(enteredValue);
+    }
+    function checkEnteredValue(enteredValue) {
```
Всё ещё вынеси выше функцию, за пределы этого `eventListener`

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,211 @@
+    }
+    function checkEnteredValue(enteredValue) {
+      if (enteredValue === '') {
+        return;
```
Желательно вернуть `return false;`

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,211 @@
+        if (checkEnteredValue(enteredValue)) {
+          clearErrorboxSheet();
+          createCheckboxItem(enteredValue);
+        } else return;
```
Лучше оставить `{}` и возвращать значение - лучше `return false;`

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,211 @@
+      }
+      for (const enteredValue of arrayEnteredValues) {
+        if (checkEnteredValue(enteredValue)) {
+          clearErrorboxSheet();
```
А зачем делать на каждое валидное значение?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,211 @@
+function getNeededAction() {
+  const controlRadioOptionCollections =
+    document.querySelectorAll(controlRadioOptionID);
+  let i = 0;
+  for (i; i < controlRadioOptionCollections.length; i++) {
+    if (controlRadioOptionCollections[i].checked) {
+      break;
+    }
+  }
```
Лучше просто найти этот index и вернуть из отдельной функции.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,211 @@
+      break;
+    }
+  }
+  const neededAction =
+    i === 0
+      ? 'addition'
+      : i === 1
+      ? 'subtraction'
+      : i === 2
+      ? 'multiplication'
+      : i === 3
+      ? 'division'
```
Всё ещё есть магические числа, нужно убрать. Такие длинные тернарные операции плохо понимаются, лучше не делать так. 

Почему бы не попробовать сделать у checkbox `value` подходящий?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,211 @@
+  return neededAction;
+}
+
+function addition(arrayEnteredValuesChecked) {
+  countingResults.innerHTML = arrayEnteredValuesChecked.reduce(
+    (addition, current) => addition + current
+  );
+}
+function subtraction(arrayEnteredValuesChecked) {
+  countingResults.innerHTML = arrayEnteredValuesChecked.reduce(
+    (subtraction, current) => subtraction - current
+  );
+}
+function multiplication(arrayEnteredValuesChecked) {
+  countingResults.innerHTML = arrayEnteredValuesChecked.reduce(
+    (multiplication, current) => multiplication * current
+  );
+}
+function division(arrayEnteredValuesChecked) {
+  if (arrayEnteredValuesChecked.includes(0)) {
+    createError('Делить на ноль нельзя!');
+    clearResultsBlock();
+  } else {
+    countingResults.innerHTML = arrayEnteredValuesChecked.reduce(
+      (division, current) => division / current
+    );
+  }
+}
```
Пусть возвращают значение после вычисления в массиве.

seal-developer committed:
1. Вынесла функции;
2.  Возвращаю return false;
3. Убрала из цикла clearErrorboxSheet();
4. Переделала определение выбранного radio;

seal-developer committed:
1. Сделала, чтобы функции вычислений  возвращали значение после вычисления в массиве.
2. Сделала, чтобы обрабатывался case с ошибкой после правок - если не выбран radio

seal-developer committed:
Поправила error для radio

seal-developer committed:
Убрала лог

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,208 @@
+    checkEnteredValues(arrayEnteredValues);
+
+    document.getElementById(enteredValuesID).value = '';
+  });
```
Унеси эту функцию вниз, так как теперь становится не очень понятно, что и как связано, пусть "подписки" и "event listener'ы" будут внизу

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,208 @@
+      return false;
+    }
+  }
+}
```
Эта функция никогда не возвращает "true" - почему?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,208 @@
+    return false;
+  }
+
+  clearErrorboxSheet();
```
Почему мы сначала добавляем ошибку, а потом чистим поле для ошибок?

Tauders replied:
Что значит `ErrorboxSheet`?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,208 @@
+      countingResults.innerHTML = multiplication(arrayEnteredValuesChecked);
+      break;
+    case 'division':
+      countingResults.innerHTML = division(arrayEnteredValuesChecked);
```
Может стоит вынести вычисление в отдельную переменную и просто потом назначить эту переменную в итог? Так, кажется, будет проще выглядеть.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,208 @@
+
+  switch (action) {
+    case 'addition':
+      countingResults.innerHTML = addition(arrayEnteredValuesChecked);
+      break;
+    case 'subtraction':
+      countingResults.innerHTML = subtraction(arrayEnteredValuesChecked);
+      break;
+    case 'multiplication':
+      countingResults.innerHTML = multiplication(arrayEnteredValuesChecked);
+      break;
+    case 'division':
+      countingResults.innerHTML = division(arrayEnteredValuesChecked);
```
А если здесь ничего не вернётся?

seal-developer replied:
Что ты имеешь ввиду? 
Если не один из кейсов не подойдет?

Tauders replied:
Нет, это только к последней операции. `division` может ничего не вернуть, судя по коду

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,208 @@
+
+  clearErrorboxSheet();
+
+  const action = determineValueSelectedRadioOption();
```
Лучше `get` вместо `determine`.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,208 @@
+    if (controlRadioOptionCollections[i].checked) {
+      return controlRadioOptionCollections[i].value;
+    }
+  }
```
Конвертируй в for..of?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,208 @@
+      (division, current) => division / current
+    );
+  }
+}
```
Лучше вынеси отображение ошибки и очистку результатов за пределы функции. В идеале функция выполняет 1 функцию и отвечает на вопрос глагола. 
У тебя же функция делает несколько вещей.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,208 @@
+  }
+
+  if (arrayEnteredValuesChecked.length === 0) {
+    clearResultsBlock();
```
А почему мы только здесь очищает результат?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,208 @@
+    return createError('Для выполнения операций выберите числа!');
+  }
+
+  clearErrorboxSheet();
```
А почему мы только здесь очищаем ошибки?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,208 @@
+    case 'error':
+      createError('Выберите операцию!');
+      clearResultsBlock();
+  }
```
Вынеси switch..case в отдельную функцию.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,208 @@
+  }
+}
+
+function performСalculationOperation() {
```
А ты уверена, что эта функция в таком виде нужна? Здесь всё в перемешку - данные берутся из документа, валидация, выполнение действий. 
Мне кажется можно это оставить в главной функции. Или разбить маленький части каждые на отдельные функции и вызывать их из главной функции.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,208 @@
+  .addEventListener('submit', function (e) {
+    e.preventDefault();
+    clearResultsBlock();
+    clearCheckboxSheet();
+    clearErrorboxSheet();
```
Что здесь значит `sheet`?

seal-developer replied:
Тут про последнюю строку, или про оба?

seal-developer replied:
Последнюю изменила, первая означает "очистка листа чекбоксов" , я не знаток английского, но гугл переводит верно

Tauders replied:
А чем `clearCheckboxList()` не подходит?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,208 @@
+  labelItem.append(enteredValue);
+}
+
+function checkEnteredValue(enteredValue) {
+  if (enteredValue === '') {
+    return false;
+  }
+  if (isNaN(enteredValue)) {
+    clearCheckboxSheet();
+    createError('Вы ввели нечисловое значение! Попробуйте снова.');
+    return false;
+  }
+  if (enteredValue.length >= 20) {
+    clearCheckboxSheet();
+    createError(
+      'Вы ввели слишком длинное число! Максимально разрешенная  длина числа - 20 знаков.'
+    );
+    return false;
+  }
+  return true;
+}
```
Почему `clearCheckboxSheet` выполняется не во всех случаях?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,208 @@
+  return true;
+}
+function checkEnteredValues(arrayEnteredValues) {
+  if (arrayEnteredValues.length >= 20) {
```
Вынеси `20` в константу (раньше я упускал что-то этот момент)

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,208 @@
+    createError('Вы ввели нечисловое значение! Попробуйте снова.');
+    return false;
+  }
+  if (enteredValue.length >= 20) {
```
Вынеси `20` в константы.

Commit:
1.Перемещение функций; 2. Корректировка checkEnteredValues(arrayEnteredValues); 3. Исправила логику чистки; 4.  Переименовала функцию создания ошибки; 5. Вычисление отдельно - присвоение отдельно; 6. Поправила случай с делением; 7. Конвертировала в for..of:
8. Вынесла все лишнее из division();
9. Вынесла switch..case в отдельную функцию;
10.  Убрала ненужную функцию - разбила все на маленькие функции;
11. Убрала магические числа в константы;

Commit:
Добавила, чтобы очищался весь лист, если одно значение не проходит проверку.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,225 @@
+      break;
+    case 'division':
+      if (checkImproverValue(arrayEnteredValuesChecked)) {
+        calculationResults = '';
```
А почему здесь `calculationResults` строка, а в других местах число? Может лучше возвращать `undefined`?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,225 @@
+
+function checkLengthArrayEnteredValuesChecked(arrayEnteredValuesChecked) {
+  if (arrayEnteredValuesChecked.length === 0) {
+    return createError('Для выполнения операций выберите числа!');
```
А зачем здесь `return`?

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,225 @@
+  }
+  if (enteredValue.length >= MAXIMUM_ALLOWED_LENGTH_NUMBER) {
+    createError(
+      'Вы ввели слишком длинное число! Максимально разрешенная  длина числа - 20 знаков.'
```
`разрешенная  длина` - 2 пробела подряд

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,225 @@
+  labelItem.append(enteredValue);
+}
+
+function checkEnteredValue(enteredValue) {
```
Можно назвать `validateEnteredValue`

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,225 @@
+  }
+
+  for (const enteredValue of arrayEnteredValues) {
+    if (checkEnteredValue(enteredValue)) {
+      createCheckboxItem(enteredValue);
+    } else {
+      clearCheckboxSheet();
+      return false;
+    }
```
Можно инвертировать и будет меньше вложенности:

```ts
    if (!checkEnteredValue(enteredValue)) {
      clearCheckboxSheet();
      return false;
    }

    createCheckboxItem(enteredValue);
```

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -0,0 +1,225 @@
+      break;
+    case 'error':
+      createError('Выберите операцию!');
+      calculationResults = '';
```
Почему строка, а обычно число? Может лучше возвращать `undefined`?

seal-developer committed:
1. Возвращаю в качестве значения calculationResults - undefined в случае ошибки;
2. Убрала ненужный return в checkLengthArrayEnteredValuesChecked(arrayEnteredValuesChecked);
3. Поправила опечатку в тексте;
4. Переименовала функцию checkEnteredValue(enteredValue) в validateEnteredValue(enteredValue);
5. Инвертировала логическое значение в цикле перебора значений массива.

Tauders commented on SubtasksJs/Nika/SubtaskJS-3/main.js:
```diff
@@ -217,9 +216,11 @@ document
-    countingResults.innerHTML = performSelectedAction(
-      arrayEnteredValuesChecked
-    );
+    if (performSelectedAction(arrayEnteredValuesChecked) != undefined) {
```
Лучше `!==` - так проверка строже. https://ru.stackoverflow.com/questions/431662/В-чем-разница-между-и

seal-developer committed:
Строгая проверка
